Objects/cown.c

```diff
@@ -0,0 +1,323 @@
+static PyObject *PyCown_set(PyCownObject *self, PyObject *args);
+static PyObject *PyCown_get(PyCownObject *self, PyObject *args);
+static PyObject *PyCown_acquire(PyCownObject *self, PyObject *args);
+
+static void PyCown_dealloc(PyCownObject *self) {
+#ifndef NDEBUG
+    fprintf(stderr, "Cown deallocated (%p)\n", self);
+#endif
+    sem_destroy(&self->semaphore);
+
+    PyTypeObject *tp = Py_TYPE(self);
+    PyObject_GC_UnTrack((PyObject *)self);
+    Py_XDECREF(self->value);
+    self->value = NULL;
+    PyObject_GC_Del(self);
+    Py_DECREF(tp);
+}
+
+static int PyCown_init(PyCownObject *self, PyObject *args, PyObject *kwds) {
+    // TODO: should not be needed in the future
+    _Py_MakeImmutable(_PyObject_CAST(Py_TYPE(self)));
+    _Py_notify_regions_in_use();
+
+#ifndef NDEBUG
+    fprintf(stderr, "Cown created (%p)\n", self);
+#endif
+
+    sem_init(&self->semaphore, 0, 0);
+    Py_SET_REGION(self, _Py_COWN);
+
+    static char *kwlist[] = {"value", NULL};
+    PyObject *value = NULL;
+
+    // See if we got a value as a keyword argument
+    if (!PyArg_ParseTupleAndKeywords(args, kwds, "|O", kwlist, &value)) {
+        return -1;  // Return -1 on failure
+    }
+
+    if (value) {
+        PyCown_set_unchecked(self, value);
+
+    } else {
+        atomic_store(&self->state, Cown_RELEASED);
+        self->value = Py_None;
+    }
+    return 0;
+}
+
+static int PyCown_traverse(PyCownObject *self, visitproc visit, void *arg) {
+    Py_VISIT(self->value);
+    return 0;
+}
+
+#define bail_if_owned(o, msg) \
+    do { \
+        /* Note: we must hold the GIL at this point -- note for future threading implementation. */ \
+        if (o->owning_thread != 0) { \
+            PyErr_Format(PyExc_RegionError, "%s: %S -- %zd", msg, o, o->owning_thread); \
+            return NULL; \
+        } \
+    } while(0);
+
+#define bail_unless_owned(o, msg) \
+    do { \
+        /* Note: we must hold the GIL at this point -- note for future threading implementation. */ \
+        PyThreadState *tstate = PyThreadState_Get(); \
+        if (o->owning_thread != tstate->thread_id) { \
+            PyErr_Format(PyExc_RegionError, "%s: %S", msg, o); \
+            return NULL; \
+        } \
+    } while(0);
+
+#define bail_unless_in_state(o, expected_state, msg) \
+    do { \
+        /* Note: we must hold the GIL at this point -- note for future threading implementation. */ \
+        if (o->state != expected_state) { \
+            PyErr_Format(PyExc_RegionError, "%s: %S", msg, o); \
+            return NULL; \
+        } \
+    } while(0);
+
+#define bail_unless_acquired(o, msg) \
+    bail_unless_owned(o, msg) \
+    bail_unless_in_state(o, Cown_ACQUIRED, msg)
+
+static PyObject *PyCown_acquire(PyCownObject *self, PyObject *args) {
+    int expected = 0;
+
+    while (!atomic_compare_exchange_strong(&self->state, &expected, Cown_ACQUIRED)) {
+        sem_wait(&self->semaphore);
+        expected = 0;
```

Comment:
I assume this is waiting for the `Cown_RELEASED` status?

```suggestion
    int expected = Cown_RELEASED;

    while (!atomic_compare_exchange_strong(&self->state, &expected, Cown_ACQUIRED)) {
        sem_wait(&self->semaphore);
        expected = Cown_RELEASED;
```

Reply:
Good catch!